solutions/cses-1734.mdx

```diff
@@ -15,61 +15,196 @@ Sort the queries in a specific order before answering them.
 
 **Time Complexity:** $\mathcal{O}(N\log{N})$
 
-Answer queries in order of the right indices. Use a BIT to keep track of indices
-that contain distinct values. For every value, update the latest index for it
-and answer all queries at the current index. This ensures later indices are
-included in the BIT to accurately answer queries.
+We answer the queries from right to left, sorted by their left indices. By using a BIT, we can keep track of indices that contain distinct values. For every value $arr[i]$, we set the latest (i.e. the left-most) index in $bit[i]$ to $1$. If this value appeared before, we set the last used index back to $0$. Then, we can answer all queries with $a == i$ by summing up the values between $[i, b]$. This ensures later indices after $i$ are included in the BIT to accurately answer queries.
 
+<LanguageSection>
+
+<CPPSection>
 ```cpp
 #include <bits/stdc++.h>
-
 using namespace std;
 
-const int MX = 4e5 + 5;
-
-int bit[MX];
-int n, q, sol[MX];
-vector<pair<int, int>> query[MX];
-vector<int> x(MX);
-map<int, int> fst;
-
-int qry (int i) {
-	int res = 0;
-	for (; i; i -= i&(-i)) {
-		res += bit[i];
+struct BIT {
+	int size;
+	vector<int> bit;
+	BIT(int n) : size(n), bit(n, 0) {}
+	void update(int x, int v) {
+		for (; x < size; x += x & (-x)) {
+			bit[x] += v;
+		}
 	}
-	return res;
-}
-
-void upd (int i, int val) {
-	for (; i <= n; i += i&(-i)) {
-		bit[i] += val;
+	/** @return sum of the values in [0,b] */
+	int query(int b) {
+		int result = 0;
+		for (; b > 0; b -= b & (-b)) {
+			result += bit[b];
+		}
+		return result;
 	}
-}
+};
 
-int main () {
+int main() {
+	int n, q;
 	cin >> n >> q;
-	for (int i = 0; i < n; i++) {
-		cin >> x[i];
+
+	// arr, queries are 1-indexed.
+	vector<int> arr(n + 1);
+	vector<vector<pair<int, int>>> queries(n + 1);
+	for (int i = 1; i <= n; i++) {
+		cin >> arr[i];
 	}
 	for (int i = 0; i < q; i++) {
 		int a, b;
 		cin >> a >> b;
-		query[a].push_back(make_pair(b, i));
+		queries[a].push_back({b, i});
 	}
-	for (int i = n; i >= 1; i--) {
-		int z = x[i - 1];
-		if (fst.count(z)) upd(fst[z], -1);
-		fst[z] = i;
-		upd(i, 1);
-		for (auto t : query[i]) {
-			sol[t.second] = qry(t.first);
+
+	// BIT also has to be 1-indexed, or the bitwise operator won't work.
+	BIT bit(n + 1);
```

Comment:
if the driver code has to do some weird crap to make sure the class works, then the class is implemented wrong
i'd put the +1 in the constructor (see the bit code i wrote for ref)